frontend/userForm.tsx

```diff
@@ -0,0 +1,306 @@
+/**
+ * @file This file contains the UserForm function. This function is used to
+ * change user information and permissions.
+ */
+import React from 'react'
+import { TextField, Checkbox, FormControlLabel,
+         Button, TableContainer, Paper, Table, TableBody, TableRow, TableCell, TableHead } from '@material-ui/core';
+import Select from 'react-select';
+import { makeStyles } from '@material-ui/core/styles';
+import { get, post, update } from './communication';
+import { useDataContext } from './data_context'
+import { Herd, Genebank } from '~data_context_global';
+import { useHistory } from 'react-router-dom';
+
+// Define styles for tab menu
+const useStyles = makeStyles({
+  form: {
+    borderLeft: "1px solid rgba(0,0,0,0.1)",
+    paddingLeft: "0.5cm",
+    display: "flex",
+    flexDirection: "column",
+  },
+  simpleField: {
+    width: "400px",
+  },
+  permissionTable: {
+  },
+  permissionTitle: {
+    fontWeight: "bold",
+    textTransform: "capitalize",
+  }
+});
+
+/** description of a user to manage */
+export interface ManagedUser {
+  id: number | 'new'
+  email: string
+  validated: boolean
+  privileges: any
+  password?: string
+}
+
+const defaultValues: ManagedUser = {
+  id: -1,
+  email: "",
+  validated: false,
+  privileges: [],
+  password: ""
+}
+
+/** These are the permission levels that can be added using the user form */
+const PermissionLevels = [{value: 'owner', label: 'Owner'},
+                          {value: 'manager', label: 'Manager'},
+                          {value: 'specialist', label: 'Genetic Specialist'},
+                          ];
+/**
+ * Provides a form for changing user metadata and user roles. The form will
+ * load data for the user id `id` if it's a number, or for a new user if `id` is
+ * `undefined` or `'new'`.
+ */
+export function UserForm({id}: {id: number | 'new' | undefined}) {
+  const {genebanks, loadData} = useDataContext()
+  const [user, setUser] = React.useState({...defaultValues} as ManagedUser)
+  const [isNew, setNew] = React.useState(false)
+  const [level, setLevel] = React.useState(PermissionLevels[0])
+  const [genebank, setGenebank] = React.useState(genebanks.length > 0 ? {value: genebanks[0].id, label: genebanks[0].name} : null)
+  const [herd, setHerd] = React.useState(null as any)
+  const history = useHistory()
+  const classes = useStyles();
+
+  const loadUser = (id: number) => {
+    get(`/api/manage/user/${id}`).then(
+      data => {
+        if (data) {
+          setUser(data)
+          setNew(false)
+        } else {
+          setNew(true)
+        }
+      },
+      error => console.error(error)
+    );
+  }
+
+  /**
+   * Loads the user information for `id`, or sets the form to accept a new user.
+   */
+  React.useEffect(() => {
+    setUser({...defaultValues})
```

Comment:
You don't need to copy the object here before setting if you maintain the invariant to never mutate the user object you get from the state (which should be maintained)

Reply:
Yeah... It just felt better to make a copy than to potentially take the risk of mutating an object that should remain static!